### PR TITLE
better handle remote files by avoiding full reads

### LINKED ISSUE #38: remote GDAL data sources are being fully downloaded
It should be possible to `geometamaker.describe` a dataset at a remote location. GDAL can do it, but if given an `https://`, it appears to download the whole file before opening it to read some header info, for a `tif`, for example. And for very large files, we observe some download errors:

```python
>>> pygeoprocessing.get_gis_type('https://storage.googleapis.com/natcap-data-cache/global/aster-v3-1s/aster-v3-1s.tif')
C:\Users\dmf\projects\invest\env\lib\site-packages\osgeo\gdal.py:312: FutureWarning: Neither gdal.UseExceptions() nor gdal.DontUseExceptions() has been explicitly called. In GDAL 4.0, exceptions will be enabled by default.
  warnings.warn(
ERROR 1: Failure writing output to destination, passed 15362 returned 0
ERROR 1: Failure writing output to destination, passed 16384 returned 0
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "C:\Users\dmf\projects\invest\env\lib\site-packages\pygeoprocessing\geoprocessing.py", line 3859, in get_gis_type
    raise ValueError(
ValueError: Could not open https://storage.googleapis.com/natcap-data-cache/global/aster-v3-1s/aster-v3-1s.tif as a gdal.OF_RASTER or gdal.OF_VECTOR.
```

Using a prefix like `/vsicurl/` seems to avoid the full download. Perhaps `geometamaker` should prepend this prefix before passing to `pygeoprocessing` and `gdal`.

Unfortunately the user cannot call `geometamaker.describe(/vsicurl/https://...)` because we also try to open the file with `frictionless` and to open any existing metadata docs with `fsspec`, which expect the plain `https://` interface.

### PR DESCRIPTION
**Pull request by davemfish**
A few metadata properties we are using required reading the entire dataset that is being described. This is okay for small, local files, but can become a problem for large, remote files. This PR updates or removes those properties so that describing a dataset remains quick.

**Properties changing:**

- `hash` -> `uid`: Instead of hashing a file to create a unique identifier, we now hash some properties of the file. I renamed to `uid` so that we do not mislead users into thinking this is a hash of the entire file.
- `rows` & `fields`: simply counting the number of rows in a table could be a problem, so this is no longer part of the metadata. Number of fields also seems unnecessary; all fields are still listed under the `schema`.
- `n_features`: a new property for vectors, instead of `rows`. It's already part of a GDAL vector metadata.

We also now use `vsicurl` when opening a remote file with GDAL. Opening with `http` seemed to download the whole file. Now geometamaker supports `http`, `https`, and local filesystem protocols, including networked drives. We could support other protocols in the future, but for now any others raise a helpful exception.

Fix #38

### CODE REVIEW

File: tests/test_geometamaker.py [hunk 1]
```diff
@@ -501,3 +501,11 @@ def test_write_to_local_workspace(self):
         self.assertTrue(
             os.path.exists(os.path.join(
                 temp_dir, f'{os.path.basename(datasource_path)}.yml')))
+
+    def test_describe_remote_datasource(self):
+        """Test describe on a file at a public url."""
+        import geometamaker
+
+        filepath = 'https://storage.googleapis.com/natcap-data-cache/global/aster-v3-1s/aster-v3-1s.tif'
```
[the remark this answers is not in the document]
Comment (phargogh) [reply tests/test_geometamaker.py:509]: I think this is a pretty reasonable thing to do, but, as you say, it expands the number of ways that the tests can fail that have nothing to do with the code in this repo.

Setting up a local HTTP server would be a good solution, and I could see another alternative being to use `mock` as well to simulate the return value of a simple file.

But for where we are in this project, I think the approach you have here would be just fine.

**Review by phargogh — CHANGES_REQUESTED** [review]
Thanks @davemfish !  I just had a couple very minor comments and suggestions for your consideration, but overall this looks great.

File: src/geometamaker/geometamaker.py [hunk 2]
```diff
@@ -13,8 +15,31 @@
...
+
+    Args:
+        filepath (str): path to a file to be opened by GDAL
+        scheme (str): the protocol prefix of the filepath
+
+    Returns:
+        str
+
+    """
+    if 'http' in scheme:
```
[the remark this answers is not in the document]
Comment (davemfish) [reply src/geometamaker/geometamaker.py:37]: Yes, that's better!

File: src/geometamaker/geometamaker.py [hunk 3]
```diff
@@ -54,7 +81,38 @@ def detect_file_type(filepath):
...
+
+    Returns:
+        dict
+
+    """
+    description = frictionless.describe(source_dataset_path).to_dict()
+    # Frictionless can retrieve stats like file size, but doing so necessarily
+    # computes a sha256 hash, which is too costly for remote or large files.
+    # So we use fsspec to get basic stats and construct a simple hash
+    of = fsspec.open(source_dataset_path)
```
[the remark this answers is not in the document]
Comment (davemfish) [reply src/geometamaker/geometamaker.py:98]: This is definitely a question I have asked myself too! Elsewhere before a call to `fsspec.open` I have this comment: `# Despite naming, this does not open a file that must be closed`, which I *think* is true, but it's confusing. `of.close()` is a  method, but I don't think it does anything unless `of.open()` was called, which I'm pretty sure it is not. `of` is an `OpenFile` instance: https://filesystem-spec.readthedocs.io/en/latest/_modules/fsspec/core.html#OpenFile

And here are some examples: https://filesystem-spec.readthedocs.io/en/latest/developer.html

Basically, we're using `fsspec.open` as a convenient way to access the filesystem (`of.fs`) and we're not actually opening the file itself. That's how I understand it anyway

File: src/geometamaker/geometamaker.py [hunk 4]
```diff
@@ -54,7 +81,38 @@ def detect_file_type(filepath):
...
+    info = of.fs.info(source_dataset_path)
+    description['bytes'] = info['size']
+    # files opened from http do not have mtime
```
[the remark this answers is not in the document]
Comment (davemfish) [reply src/geometamaker/geometamaker.py:101]: Thanks for the suggestion, I dropped the use of `fsspec` in this instance as I do think it's worth getting the modified time.

File: tests/test_geometamaker.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (davemfish) [reply tests/test_geometamaker.py:509]: Alright I'm going to leave it as-is for now. Thanks for setting up that `natcap-data-cache`! It feels like a place we can depend on for a while.

**Review by phargogh — APPROVED** [review]
Looks great to me, thanks @davemfish !
